server/aesdsocket.c

```diff
@@ -0,0 +1,423 @@
+        goto error_1;
+    }
+    rc = sigaction(SIGTERM, &signal_action, 0);
+    if (rc < 0 ) {
+        syslog(LOG_ERR, "aesdsocket: Sigaction failed for SIGTERM %s", strerror(errno));
+        goto error_1;
+    }
+
+    soc_server = socket(AF_INET, SOCK_STREAM, 0);
+    if (soc_server < 0) {
+        rc = -1;
+        syslog(LOG_ERR, "aesdsocket: Socket creation failed %s", strerror(errno));
+        goto error_1;
+    }
+
+    rc = setsockopt(soc_server, SOL_SOCKET, SO_REUSEPORT, &cmd_option, sizeof(cmd_option));
+    if (rc < 0) {
+        syslog(LOG_ERR, "aesdsocket: API setsockopt failed %s", strerror(errno));
+        goto exit_shutdown;
+    }
+
+    aesdsoc_addr.sin_family = AF_INET;
+    aesdsoc_addr.sin_addr.s_addr = INADDR_ANY;
+    aesdsoc_addr.sin_port = htons(SOCKET_PORT);
+
+    rc = bind(soc_server, (struct sockaddr*)&aesdsoc_addr, aesdsoc_addr_len);
+    if (rc < 0) {
+        syslog(LOG_ERR, "aesdsocket: API bind failure %s", strerror(errno));
+        goto exit_shutdown;
+    }
+
+    while( exit_aesd_soc == FALSE) {
+        rc = listen(soc_server, MAX_SERVER_CONNECTION);
+        if (rc < 0) {
+            syslog(LOG_ERR, "aesdsocket: API listen failed %s", strerror(errno));
+            goto exit_shutdown;
+        }
+
+        soc_client = accept(soc_server, (struct sockaddr*)&aesdsoc_addr, 
+            (socklen_t*)&aesdsoc_addr_len);
+        if (soc_client < 0 ) {
+            rc = -1;
+            syslog(LOG_ERR, "aesdsocket: API accept failed %s", strerror(errno));
+            goto exit_shutdown;
+        }
+
+        syslog(LOG_INFO, "Accepted connection from %s", inet_ntoa(aesdsoc_addr.sin_addr)); 
+
+                
+        rc = fcntl(soc_client, F_SETFL, O_NONBLOCK);
+        if (rc < 0) {
+            syslog(LOG_ERR, "aesdsocket: fcntl failed %s", strerror(errno));
+            goto exit_client;
+        }
+        rc = clock_gettime(CLOCK_MONOTONIC, &start);
+        if (rc < 0) {
+            syslog(LOG_ERR, "aesdsocket: clock_gettime failed %s", strerror(errno));
+            goto exit_client;
+        }
+
+        int total_bytes_received = 0;
+        buffer = (char*)malloc(buffer_size);
+        if(buffer == NULL) {    
+            rc = -1;
+            syslog(LOG_ERR, "aesdsocket: malloc failed %s", strerror(errno));
+            goto exit_client;
+        }
+        while(1) {
+            rcv_data_len = recv(soc_client, buffer + total_bytes_received, buffer_size - total_bytes_received, 0);
+            if (rcv_data_len < 0) {
+                rc = clock_gettime(CLOCK_MONOTONIC, &now);
+                if (rc < 0) {            
+                    syslog(LOG_ERR, "aesdsocket: clock_gettime failed %s", strerror(errno));
+                    goto goto_mem_cleanup;
+                }
+                diff_time = ((now.tv_sec - start.tv_sec) * 1000000) + ((now.tv_nsec - start.tv_nsec) / 1000); 
+                if(diff_time > MSEC_2_USEC(PACKET_TIMEOUT_END)) {   
```

Comment:
@sujoyray-ucolorado I don't think it's safe to assume that a packet is complete after 10ms between redcv calls.  I think you need to look for the end of packet `\n`

Reply:
Hi Dan, \n is checked in a separate function. This is a timeout to cover - if client is sending a broken packet say "12345" ..... long wait then "6789\n".

Reply:
@dwalkes : Please have a look at **process_and_save_data**, where \n has been checked before saving the file

Reply:
Is this trying to handle partial packets which weren't terminated with `\n` after a timeout?  If so, you don't need to handle this, and it might cause problems to try to handle this.